### DISCUSSION TITLE
The first beta release of NetBox v3.7 is available!

### DISCUSSION THREAD

**Discussion by jeremystretch** [netbox-community/netbox#14430] [Announcements]
:warning: **This is a beta release** intended only for testing and evaluation. Do not deploy it in a production environment. No upgrade path to future releases will be provided.

### Breaking Changes

* The following fields have been removed from the Webhook model: `content_types`, `type_create`, `type_update`, `type_delete`, `type_job_start`, `type_job_end`, `enabled`, and `conditions`. Webhooks are now tied to events via [event rules](https://docs.netbox.dev/en/feature/features/event-rules). Existing webhooks will have event rules created automatically upon upgrade.
* The `ui_visibility` field on the [custom field model](https://docs.netbox.dev/en/feature/models/extras/customfield/) has been replaced with two new fields: `ui_visible` and `ui_editable`. Existing values will be migrated automatically upon upgrade.
* The `FeatureQuery` class for querying content types by model feature has been removed. Plugins should now use the new `with_feature()` manager method on NetBox's proxy model for ContentType.
* The ConfigRevision model has been moved from `extras` to `core`. Configuration history will be retained throughout the upgrade process.
* The L2VPN and L2VPNTermination models have been moved from the `ipam` app to the new `vpn` app. All object data will be retained, however please note that the relevant API endpoints have moved to `/api/vpn/`.
* The `CustomFieldsMixin`, `SavedFiltersMixin`, and `TagsMixin` classes have moved from the `extras.forms.mixins` module to `netbox.forms.mixins`.

#### VPN Tunnels ([#9816](https://github.com/netbox-community/netbox/issues/9816))

Several new models have been introduced to enable [VPN tunnel management](https://docs.netbox.dev/en/feature/features/vpn-tunnels/). Users can now define tunnels with two or more terminations to replicate peer-to-peer or hub-and-spoke topologies. Each termination is made to a virtual interface on a device or VM. Additionally, users can define IKE and IPSec policies which can be applied to tunnels to document encryption and authentication strategies.

#### Event Rules ([#14132](https://github.com/netbox-community/netbox/issues/14132))

This release introduces [event rules](https://docs.netbox.dev/en/feature/features/event-rules/), which can be used to send webhooks or execute custom scripts automatically in response to NetBox events. For example, it's now possible to run a custom script whenever a new site is created with a particular status or tag.

Event rules replace and extend functionality that was previously built into the webhook model. Event rules will be created for any existing webhooks upon upgrade.

#### Virtual Machine Disks ([#8356](https://github.com/netbox-community/netbox/issues/8356))

A new [VirtualDisk](https://docs.netbox.dev/en/feature/models/virtualization/virtualdisk/) model has been introduced to enable tracking the assignment of discrete virtual disks to virtual machines. The original `size` field has been retained on the VirtualMachine model, and will be automatically updated with the aggregate size of all assigned virtual disks. (Users who opt to eschew the new model may continue using the VirtualMachine `size` attribute as before.)

#### Object Protection Rules ([#10244](https://github.com/netbox-community/netbox/issues/10244))

A new [`PROTECTION_RULES`](https://docs.netbox.dev/en/feature/configuration/data-validation/#protection_rules) configuration parameter is now available. Similar to how [custom validation rules](https://docs.netbox.dev/en/feature/customization/custom-validation/) can be used to enforce certain values for object attributes, protection rules guard against the deletion of objects which do not meet specified criteria. This enables an administrator to prevent, for example, the deletion of a site which has a status of "active."

#### Improved Custom Field Visibility Controls ([#13299](https://github.com/netbox-community/netbox/issues/13299))

The old `ui_visible` field on the custom field model](https://docs.netbox.dev/en/feature/models/extras/customfield/) has been replaced by two new fields, `ui_visible` and `ui_editable`, which control how and whether a custom field is displayed when view and editing an object, respectively. Separating these two functions into discrete fields enables more control over how each custom field is presented to users. The values of these fields will be appropriately set automatically during the upgrade process depending on the value of the original field.

#### Improved Global Search Results ([#14134](https://github.com/netbox-community/netbox/issues/14134))

Global search results now include additional context about each object, such as a description, status, and/or related objects. The set of attributes to be displayed is specific to each object type, and is defined by setting `display_attrs` under the object's [SearchIndex class](https://docs.netbox.dev/en/feature/plugins/development/search/#netbox.search.SearchIndex).

#### Table Column Registration for Plugins ([#14173](https://github.com/netbox-community/netbox/issues/14173))

Plugins can now [register their own custom columns](https://docs.netbox.dev/en/feature/plugins/development/tables/#extending-core-tables) for inclusion on core NetBox tables. For example, a plugin can register a new column on SiteTable using the new `register_table_column()` utility function, and it will become available for users to select for display.

#### Data Backend Registration for Plugins ([#13381](https://github.com/netbox-community/netbox/issues/13381))

Plugins can now [register their own data backends](https://docs.netbox.dev/en/feature/plugins/development/data-backends/) for use with [synchronized data sources](https://docs.netbox.dev/en/feature/features/synchronized-data/). This enables plugins to introduce new backends in addition to the git, S3, and local path backends provided natively.

### Enhancements

* [#12135](https://github.com/netbox-community/netbox/issues/12135) - Avoid orphaned interfaces by preventing the deletion of interfaces which have children assigned
* [#12216](https://github.com/netbox-community/netbox/issues/12216) - Add a `color` field for circuit types
* [#13230](https://github.com/netbox-community/netbox/issues/13230) - Allow device types to be excluded from consideration when calculating a rack's utilization
* [#13334](https://github.com/netbox-community/netbox/issues/13334) - Added an `error` field to the Job model to record any errors associated with its execution
* [#13427](https://github.com/netbox-community/netbox/issues/13427) - Introduced a mechanism for omitting models from general-purpose lists of object types
* [#13690](https://github.com/netbox-community/netbox/issues/13690) - Display any dependent objects to be deleted prior to deleting an object via the web UI
* [#13794](https://github.com/netbox-community/netbox/issues/13794) - Any models with a relationship to Tenant are now included automatically in the list of related objects under the tenant view
* [#13808](https://github.com/netbox-community/netbox/issues/13808) - Added a `/render-config` REST API endpoint for virtual machines
* [#14035](https://github.com/netbox-community/netbox/issues/14035) - Order objects of equivalent weight by value in global search results to improve readability
* [#14156](https://github.com/netbox-community/netbox/issues/14156) - Enable custom fields for contact assignments
* [#14361](https://github.com/netbox-community/netbox/issues/14361) - Add a `description` field for webhooks
* [#14365](https://github.com/netbox-community/netbox/issues/14365) - Introduced `job_start` and `job_end` signals

### Other Changes

* [#13550](https://github.com/netbox-community/netbox/issues/13550) - Optimized the format for declaring view actions under `ActionsMixin` (backward compatibility has been retained)
* [#13645](https://github.com/netbox-community/netbox/issues/13645) - Installation of the `sentry-sdk` Python library is now required only if Sentry reporting is enabled
* [#14036](https://github.com/netbox-community/netbox/issues/14036) - Move plugin resources from the `extras` app into `netbox` (backward compatibility has been retained)
* [#14153](https://github.com/netbox-community/netbox/issues/14153) - Replace `FeatureQuery` with new `with_feature()` method on ContentType manager
* [#14311](https://github.com/netbox-community/netbox/issues/14311) - Move the L2VPN models from the `ipam` app to the new `vpn` app
* [#14312](https://github.com/netbox-community/netbox/issues/14312) - Move the ConfigRevision model from the `extras` app to `core`
* [#14326](https://github.com/netbox-community/netbox/issues/14326) - Form feature mixin classes have been moved from the `extras` app to `netbox`
* [#14395](https://github.com/netbox-community/netbox/issues/14395) - Moved `extras.webhooks_worker.process_webhook()` to `extras.webhooks.send_webhook()` (backward compatibility has been retained)

<hr /><em>This discussion was created from the release <a href='https://github.com/netbox-community/netbox/releases/tag/v3.7-beta1'>v3.7-beta1 - 2023-12-05</a>.</em>

**Comment by iunderwood** [comment]
The note of the VPN thing makes me incredibly happy!

**Comment by markkuleinio** [comment]
Great!

Opened https://github.com/netbox-community/pynetbox/issues/595 so that I can submit the PR to add the `vpn` app in `pynetbox`.

**Comment by markkuleinio** [comment]
The new search looks promising! (Disclaimer: I have skipped 3.4-3.6 and going to upgrade from 3.3 to 3.7 so I never fought with the global search problems)

Is the searchindex class link somehow faulty? I only get this kind of page:

(deleted the massive distracting image now that the link was fixed)

**Reply by jeremystretch** [reply]
I don't think we have a model documentation page for `SearchIndex` (see the [plugin docs](https://docs.netbox.dev/en/stable/plugins/development/search/)). What link did you follow to get there?

**Reply by markkuleinio** [reply]
The one shown above:

> Global search results now include additional context about each object, such as a description, status, and/or related objects. The set of attributes to be displayed is specific to each object type, and is defined by setting display_attrs under the object's [SearchIndex class](https://docs.netbox.dev/en/feature/plugins/development/search.md#netbox.search.SearchIndex).

**Reply by jeremystretch** [reply]
Oh, it's the `.md` suffix. I'll fix it, thanks.

**Reply by markkuleinio** [reply]
Right, thanks. While you are here :-) can you point me in which file are the displayed fields (for search results) defined for devices? I'm testing the search and wishing to see the device status in the search results (going to test it with my data).

**Reply by jeremystretch** [reply]
These are defined under the `DeviceIndex` class in `dcim/search.py`.

**Reply by markkuleinio** [reply]
Oh actually had wrong branch. So https://github.com/netbox-community/netbox/blob/feature/netbox/dcim/search.py has it:

```
    display_attrs = (
        'site', 'location', 'rack', 'device_type', 'role', 'tenant', 'platform', 'serial', 'asset_tag', 'description',
    )
```

Would it be beneficial for me to open feature request to add status there as well?

Also, how (im)possible would it be to make that user-configurable in configuration.py? Asking because in one of my major setups we don't use device types (they are just "routers" so "`Device Type: Router`" is unnecessary for us. (Not a major thing but anyway)

**Reply by jeremystretch** [reply]
> Would it be beneficial for me to open feature request to add status there as well?

Sure.

> Also, how (im)possible would it be to make that user-configurable in configuration.py?

It would probably be more effort than it's worth. As with table columns, inevitably you're going to have different users with different preferences, so a single global configuration is going to be a of limited value. It might be worth a FR but probably not something we'd tackle before the v3.7.0 release.

To be honest I'm still not 100% sold on the current implementation. It still feels a bit kludgy.

**Reply by markkuleinio** [reply]
#14503 opened for the status field.

**Comment by v0tti** [comment]
I had some problem with a custom script that imported the L2VPN model like this `from ipam.models import IPAddress, Prefix, L2VPN`. I did not see any error besides NetBox not being able to import the script until I increased the logging level to debug. It's easy to overlook errors like this if not reported clearly so maybe they should be reported on a lower log level?

**Comment by markkuleinio** [comment]
I have successfully started the evaluation of 3.7-beta1 with a customer, we are really trying to get onboard with 3.7 release (from 3.3). Unfortunately the year-end busyness affected/affects that, hopefully I'm able to gather some more feedback (in addition to the already submitted issues/FRs) in the first week of January 👍

**Comment by seismiccollision** [comment]
i thought there was gonna be an online demo of this to test before the master release as i'd seen that happen for at least one beta version in the past at https://demo-beta.netbox.dev/. if that had been made available i would have definitely tried testing and evaluating at least the improved global search results in advance


however i am testing now and found this device by putting search through its paces https://demo.netbox.dev/dcim/devices/82/ (for instance and as user "admin") and the top level tabs for the device aren't exposing front-ports or rear-ports to me even though if i see ports in the global search for it and can hit url https://demo.netbox.dev/dcim/devices/82/front-ports/.
other than that, global search presentation is much improved even with attribute bubbles all as the same color, thanks a bunch!

**Reply by abhi1693** [reply]
Beta demo was made available during the entire time.

**Reply by candlerb** [reply]
Did you open an issue for the missing tabs? It does seem pretty broken.

**Reply by jcollie** [reply]
Doesn't seem to be an issue on my instance that I just upgraded to 3.7.0...

![image](https://github.com/netbox-community/netbox/assets/740022/ad459690-be58-4ea8-b437-99ef59ba5027)

**Reply by candlerb** [reply]
Interesting - I assume you've checked that the bottom right of the screen shows "(v3.7.0)". I wonder what's up with demo.netbox.dev then?

**Reply by markkuleinio** [reply]
3.7.0 ports tabs are shown correctly on my updated instance as well.

**Reply by jeremystretch** [reply]
This is an issue with the snapshots used to automatically rebuild the demo instance; it's not a bug in NetBox itself.

**Reply by seismiccollision** [reply]
thanks! where was the beta demo? i clearly missed it

**Reply by jeremystretch** [reply]
Same place it always is, `beta-demo.netbox.dev` (taken down now).

**Comment by jcollie** [comment]
> Interesting - I assume you've checked that the bottom right of the screen shows "(v3.7.0)". I wonder what's up with demo.netbox.dev then?

Yes, my version is ```netbox-5bfbc45487-c698r (v3.7.0)```
